src/exchange/exchange.py

```diff
@@ -158,16 +167,144 @@ def add_tool_use(self, tool_use: ToolUse) -> None:
+                )
+            )
+        self.checkpoint_data.checkpoints.append(
             Checkpoint(
-                start_index=(0 if not self.checkpoints else self.checkpoints[-1].end_index),
-                end_index=len(self.messages),
-                token_count=(
-                    usage.total_tokens
-                    if not self.checkpoints
-                    else usage.total_tokens - sum(cp.token_count for cp in self.checkpoints)
-                ),
-                latest_generated_tokens=usage.output_tokens,
+                start_index=len(self.messages) - 1 + self.checkpoint_data.message_index_offset,
+                end_index=len(self.messages) - 1 + self.checkpoint_data.message_index_offset,
+                token_count=second_block_token_count,
             )
         )
+
+        # TODO: check if the front of the checkpoints doesn't overlap with
+        # the first message. if so, we are missing checkpoint data from
+        # message[0] to message[checkpoint_data.checkpoints[0].start_index]
+        # we can fill in this data by performing an extra request and doing some math
+        self.checkpoint_data.total_token_count = usage.total_tokens
+
+    def pop_last_message(self) -> Message:
+        """Pop the last message from the exchange, handling checkpoints correctly"""
+        if (
+            len(self.checkpoint_data.checkpoints) > 0
+            and self.checkpoint_data.last_message_index > len(self.messages) - 1
+        ):
+            raise ValueError("Our checkpoint data is out of sync with our message data")
+        if (
+            len(self.checkpoint_data.checkpoints) > 0
+            and self.checkpoint_data.last_message_index == len(self.messages) - 1
+        ):
+            # remove the last checkpoint, because we no longer know the token count of it's contents.
+            # note that this is not the same as reverting to the last checkpoint, as we want to
+            # keep the messages from the last checkpoint. they will have a new checkpoint created for
+            # them when we call generate() again
+            self.checkpoint_data.pop()
+        self.messages.pop()
+
+    def pop_first_message(self) -> Message:
+        """Pop the first message from the exchange, handling checkpoints correctly"""
+        if len(self.messages) == 0:
+            raise ValueError("There are no messages to pop")
+        if len(self.checkpoint_data.checkpoints) == 0:
+            raise ValueError("There must be at least one checkpoint to pop the first message")
+
+        # get the start and end indexes of the first checkpoint, use these to remove message
+        first_checkpoint = self.checkpoint_data.checkpoints[0]
+        first_checkpoint_start_index = first_checkpoint.start_index - self.checkpoint_data.message_index_offset
+
+        # check if the first message is part of the first checkpoint
+        if first_checkpoint_start_index == 0:
```

Comment:
Is it so that if this is not the case, we've goofed somewhere and are not in a good state? Is there a way we could get here (I don't think so), but maybe we add a check for that to catch it

Reply:
We can totally reach both `first_checkpoint_start_index == 0` and `first_checkpoint_start_index != 0`. Just depicted an example below:

```
note: [] denotes list of messages, () denotes how checkpoints are grouped

if we start off with:
[(1,2,3),(4),(5)]

pop_first_message()
> first_checkpoint_start_index == 0
> and the resulting array: [2,3,(4),(5)]

pop_first_message() again
> first_checkpoint_start_index == 2
> and the resulting array: [3,(4),(5)]
```